src/main/java/com/epam/aidial/core/upstream/UpstreamRoute.java

```diff
@@ -6,25 +6,26 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
-import java.util.concurrent.atomic.AtomicLong;
 
 @RequiredArgsConstructor
 public class UpstreamRoute implements Iterator<Upstream> {
 
     private final List<Upstream> upstreams;
-    private final AtomicLong counter;
     private final int offset;
+    private final int maxAttempts;
 
-    private Upstream current;
-    private int count;
+    private Upstream upstream;
+    private int attempts;
+    private int next;
+    private int prev;
 
     public int attempts() {
-        return count;
+        return attempts;
```

Comment:
the field accumulates all attempts made on all upstreams.
I think we should track attempts per upstream like:
```
  return attempts[(count + offset) % upstreams.size()]
```

Reply:
We can track, but why?

Reply:
I was thinking about the concept we want to track the number of retries per upstream but not in total.